src/auth/Crypto.cc

```diff
@@ -348,24 +361,797 @@ int CryptoAES::validate_secret(const bufferptr& secret)
+          reinterpret_cast<const byte*>(secret.c_str()), CryptoAES_256_CBC::KEY_SIZE,
+          reinterpret_cast<const byte*>(iv.c_str()), CryptoAES_256_CBC::IV_SIZE);
+      d.ProcessData((byte*)out, (byte*)in, size);
+    }
+    return true;
+  }
+};
+
+#else
+#error Must define USE_CRYPTOPP or USE_NSS
+#endif
+
+CryptoKeyHandler *CryptoAES_256_CBC::get_key_handler(
+    const bufferptr& secret, const bufferptr& iv, string& error)
+{
+  if (secret.length() != KEY_SIZE) {
+    error = "Improper key size.";
+    return nullptr;
+  }
+  if (iv.length() != IV_SIZE) {
+    error = "Improper iv size.";
+    return nullptr;
+  }
+  CryptoAES_256_CBCKeyHandler *ckh = new CryptoAES_256_CBCKeyHandler();
+  if (!ckh->setup(secret, iv, error)) {
     delete ckh;
-    return NULL;
+    ckh = nullptr;
   }
   return ckh;
 }
 
+// --
+
+#ifdef USE_NSS
+
+class CryptoAES_256_ECBKeyHandler : public CryptoKeyHandler {
+public:
+  ~CryptoAES_256_ECBKeyHandler() {
+    cleanup();
+  }
+
+  int encrypt(const bufferlist& in, bufferlist& out, std::string *error) const {
+    return ecb_transform(in, out, true, error);
+  }
 
+  int decrypt(const bufferlist& in, bufferlist& out, std::string *error) const {
+    return ecb_transform(in, out, false, error);
+  }
 
+  bool setup(const bufferptr& key, std::string& error) {
+    bool result = false;
+    SECItem keyItem;
+    slot = PK11_GetBestSlot(CKM_AES_ECB, NULL);
+    if (slot) {
+      keyItem.type = siBuffer;
+      keyItem.data = const_cast<uint8_t*>(reinterpret_cast<const uint8_t*>(key.c_str()));
+      keyItem.len = CryptoAES_256_ECB::KEY_SIZE;
+
+      symkey = PK11_ImportSymKey(slot, CKM_AES_ECB, PK11_OriginUnwrap,
+                                 CKA_UNWRAP, &keyItem, NULL);
+      if (symkey) {
+        param = PK11_ParamFromIV(CKM_AES_ECB, NULL);
+        if (param) {
+          result = true;
+        }
+      }
+    }
+    if (!result) {
+      cleanup();
+      std::ostringstream ss;
+      ss << "Failed to perform AES-CBC encryption: " << PR_GetError();
+      error = ss.str();
+    }
+    return result;
+  }
+
+private:
+  PK11SlotInfo *slot = nullptr;
+  PK11SymKey *symkey = nullptr;
+  SECItem *param = nullptr;
+  PK11Context *ectx = nullptr;
+
+  void cleanup() {
+    if (param) {
+      SECITEM_FreeItem(param, PR_TRUE);
+      param = nullptr;
+    }
+    if (symkey) {
+      PK11_FreeSymKey(symkey);
+      symkey = nullptr;
+    }
+    if (slot) {
+      PK11_FreeSlot(slot);
+      slot = nullptr;
+    }
+  }
+
+  int ecb_transform(
+      const bufferlist& in,
+      bufferlist& out,
+      bool do_encrypt,
+      std::string *error) const
+  {
+    int result = -EINVAL;
+    if ((in.length() % CryptoAES_256_ECB::BLOCK_SIZE) != 0) {
+      if (error) {
+        *error = "Input size is not multiple of 16";
+      }
+    } else {
+      bufferptr p(in.length());
+      bufferlist _in(in);
+      const unsigned char* data_in = reinterpret_cast<const unsigned char*>(_in.c_str());
```

Comment:
still calling `bufferlist::c_str()` here as well - can you call `PK11_CipherOp()` with each buffer segment instead?